crates/interpreter/src/valid.rs

```diff
@@ -26,16 +26,135 @@ use crate::toctou::*;
 use crate::util::*;
 use crate::*;
 
-/// Checks whether a WASM module in binary format is valid.
-pub fn validate(binary: &[u8]) -> Result<Vec<MetadataEntry>, Error> {
+/// Checks whether a WASM module in binary format is valid, and returns the side table.
+pub fn prepare(binary: &[u8]) -> Result<Vec<MetadataEntry>, Error> {
     Context::<Prepare>::default().check_module(&mut Parser::new(binary))
 }
 
-pub trait ValidMode: Default {}
+#[allow(dead_code)]
+#[allow(unused_variables)]
+/// Checks whether a WASM module with the side table in binary format is valid.
+pub fn verify(binary: &[u8]) -> Result<(), Error> {
+    todo!()
+}
+
+trait ValidMode: Default {
+    /// List of source branches, when preparing.
+    ///
+    /// When verifying, contains at most one _target_ branch. Source branches are eagerly patched to
+    /// their target branch using the branch table.
+    type Branches<'m>: BranchesApi<'m>;
+
+    /// Branch table to prepare or verify.
+    type BranchTable<'m>: BranchTableApi<'m>;
+}
+
+trait BranchesApi<'m>: Default + IntoIterator<Item = SideTableBranch<'m>> {
+    // fn push(&mut self, branch: SideTableBranch<'m>) -> CheckResult;
+    /// Pushes a source branch, when preparing.
+    ///
+    /// When verifying, only push if there are no branches already. If there is one, verify that
+    /// it's the same.
+    fn push_branch(&mut self, branch: SideTableBranch<'m>) -> CheckResult;
+}
+
+trait BranchTableApi<'m>: Default {
+    /// Updates the branch table for source according to target, when preparing.
+    ///
+    /// When verifying, makes sure source is target branch.
```

Comment:
Actually to simplify review, and probably also your work, we could simply have those documentation as TODOs in the respective implementations. This way it's local. And we don't need documentation for those private traits, we can directly look at the implementation which will usually be shorter than the documentation.